app/src/main/java/org/mozilla/fenix/wallpapers/WallpaperMetadataFetcher.kt

```diff
@@ -0,0 +1,95 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package org.mozilla.fenix.wallpapers
+
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.withContext
+import mozilla.components.concept.fetch.Client
+import mozilla.components.concept.fetch.Request
+import org.json.JSONArray
+import org.json.JSONObject
+import org.mozilla.fenix.BuildConfig
+import java.text.SimpleDateFormat
+import java.util.Date
+import java.util.Locale
+
+/**
+ * Utility class for downloading wallpaper metadata from the remote server.
+ *
+ * @property client The client that will be used to fetch metadata.
+ */
+class WallpaperMetadataFetcher(
+    private val client: Client
+) {
+    private val metadataUrl = BuildConfig.WALLPAPER_URL.substringBefore("android") +
+        "metadata/v$currentJsonVersion/wallpapers.json"
+
+    /**
+     * Downloads the list of wallpapers from the remote source. Failures will return an empty list.
+     */
+    suspend fun downloadWallpaperList(): List<Wallpaper> = withContext(Dispatchers.IO) {
+        Result.runCatching {
+            val request = Request(url = metadataUrl, method = Request.Method.GET)
+            val response = client.fetch(request)
+            response.body.useBufferedReader {
+                val json = it.readText()
+                JSONObject(json).parseAsWallpapers()
+            }
+        }.getOrElse { listOf() }
+    }
+
+    private fun JSONObject.parseAsWallpapers(): List<Wallpaper> = with(getJSONArray("collections")) {
+        (0 until length()).map { index ->
+            getJSONObject(index).toCollectionOfWallpapers()
+        }.flatten()
+    }
+
+    private fun JSONObject.toCollectionOfWallpapers(): List<Wallpaper> {
+        val collectionId = getString("id")
+        val availableLocales = optJSONArray("available-locales")?.getAvailableLocales()
+        val availabilityRange = optJSONObject("availability-range")?.getAvailabilityRange()
+        val learnMoreUrl = optString("learn-more-url")
+        val collection = Wallpaper.Collection(
+            name = collectionId,
+            availableLocales = availableLocales,
+            startDate = availabilityRange?.first,
+            endDate = availabilityRange?.second,
+            learnMoreUrl = learnMoreUrl,
+        )
+        return getJSONArray("wallpapers").toWallpaperList(collection)
+    }
+
+    private fun JSONArray.getAvailableLocales(): List<String>? =
+        (0 until length()).map { getString(it) }
+
+    private fun JSONObject.getAvailabilityRange(): Pair<Date, Date>? {
+        val formatter = SimpleDateFormat("yyyy-MM-dd", Locale.US)
+        return formatter.parse(getString("start"))!! to formatter.parse(getString("end"))!!
```

Comment:
Could we try to avoid using `!!` as we could cause NPE? :)

Reply:
Yeah, I can update it here. Since the top-level function is wrapped in a `Result.runCatching`, throwing NPEs was actually by design in order to quickly cause failures. Looking back at it, I think it's bad to have that behavior implied and spread over the two sites, and runs a risk of causing errors if the `Result.runCatching` was ever removed.

Reply:
Thanks for the clarification, I missed the `runCatching`